### ISSUE TITLE
'Single-channel predictions not supported'

### ISSUE THREAD

**Issue by jscodecode** [JiYuanFeng/MCTrans#25]
Hello, i have a"Single channel image training error" problems。How can I solve that? Could I get your some suggestions？ Thank you very much！！！

**Comment by JiYuanFeng** [comment]
hi, can you please provide the whole log?

**Comment by jscodecode** [comment]
I tried to use our own data set. Our dataset is a dichotomous task. The data format is PNG. （1* 1* 512* 512） What do we need to pay attention to, such as the number of categories in "PanNukeDataset"?

**Comment by jscodecode** [comment]
I tried to use our own data set. Our dataset is a dichotomous task. The data format is PNG. （1*1*512*512） What do we need to pay attention to, such as the number of categories in "PanNukeDataset"?

**Comment by jscodecode** [comment]
> hi, can you please provide the whole log?

2022-09-23 20:59:37,843 - MCTrans - INFO - Distributed training: False
2022-09-23 20:59:38,286 - MCTrans - INFO - Config:
model = dict(
    type='EncoderDecoder',
    pretrained=None,
    encoder=dict(type='VGG', in_channel=512, init_channels=32, num_blocks=2),
    center=dict(
        type='MCTrans',
        d_model=128,
        nhead=8,
        d_ffn=512,
        dropout=0.1,
        act='relu',
        n_levels=3,
        n_points=4,
        n_sa_layers=6),
    decoder=dict(type='UNetDecoder', in_channels=[32, 64, 128, 128, 128]),
    seg_head=dict(
        type='BasicSegHead',
        in_channels=32,
        num_classes=6,
        post_trans=[
            dict(type='Activations', softmax=True),
            dict(type='AsDiscrete', argmax=True)
        ],
        losses=[
            dict(type='DiceLoss', softmax=True, to_onehot_y=True),
            dict(type='CrossEntropyLoss'),
            dict(type='FLoss')
        ]),
    aux_head=dict(
        type='MCTransAuxHead',
        d_model=128,
        d_ffn=512,
        act='relu',
        num_classes=6,
        in_channles=[32, 64, 128, 128, 128],
        losses=[dict(type='MCTransAuxLoss', sigmoid=True, loss_weight=0.1)]))
dataset_type = 'PanNukeDataset'
patch_size = [512, 512]
keyword = ['img', 'seg_label']
train_transforms = [
    dict(
        type='LoadImage', keys=['img', 'seg_label'], meta_key_postfix='metas'),
    dict(type='AsChannelFirst', keys='img'),
    dict(type='AddChannel', keys='seg_label'),
    dict(
        type='Resize',
        keys=['img', 'seg_label'],
        spatial_size=[512, 512],
        mode=('bilinear', 'nearest')),
    dict(type='ScaleIntensity', keys='img'),
    dict(type='RandMirror'),
    dict(type='ToTensor', keys=['img', 'seg_label'])
]
test_transforms = [
    dict(
        type='LoadImage', keys=['img', 'seg_label'], meta_key_postfix='metas'),
    dict(type='AsChannelFirst', keys='img'),
    dict(type='AddChannel', keys='seg_label'),
    dict(
        type='Resize',
        keys=['img', 'seg_label'],
        spatial_size=[512, 512],
        mode=('bilinear', 'nearest')),
    dict(type='ScaleIntensity', keys='img'),
    dict(type='ToTensor', keys=['img', 'seg_label'])
]
data = dict(
    samples_per_gpu=1,
    workers_per_gpu=1,
    train=dict(
        type='PanNukeDataset',
        img_dirs=[
            'data/medical/pannuke/split-images/BAO SU RI TA LA TU_data',
            'data/medical/pannuke/split-images/DANG YA QING_data'
        ],
        img_suffix='.png',
        label_dirs=[
            'data/medical/pannuke/split-masks/BAO SU RI TA LA TU_label',
            'data/medical/pannuke/split-masks/DANG YA QING_label'
        ],
        label_suffix='-gt.png',
        phase='train',
        transforms=[
            dict(
                type='LoadImage',
                keys=['img', 'seg_label'],
                meta_key_postfix='metas'),
            dict(type='AsChannelFirst', keys='img'),
            dict(type='AddChannel', keys='seg_label'),
            dict(
                type='Resize',
                keys=['img', 'seg_label'],
                spatial_size=[512, 512],
                mode=('bilinear', 'nearest')),
            dict(type='ScaleIntensity', keys='img'),
            dict(type='RandMirror'),
            dict(type='ToTensor', keys=['img', 'seg_label'])
        ]),
    val=dict(
        type='PanNukeDataset',
        img_dirs=['data/medical/pannuke/split-images/BAO SU RI TA LA TU_data'],
        img_suffix='.png',
        label_dirs=[
            'data/medical/pannuke/split-masks/BAO SU RI TA LA TU_label'
        ],
        label_suffix='-gt.png',
        phase='val',
        transforms=[
            dict(
                type='LoadImage',
                keys=['img', 'seg_label'],
                meta_key_postfix='metas'),
            dict(type='AsChannelFirst', keys='img'),
            dict(type='AddChannel', keys='seg_label'),
            dict(
                type='Resize',
                keys=['img', 'seg_label'],
                spatial_size=[512, 512],
                mode=('bilinear', 'nearest')),
            dict(type='ScaleIntensity', keys='img'),
            dict(type='ToTensor', keys=['img', 'seg_label'])
        ]))
log_config = dict(
    interval=10,
    hooks=[
        dict(type='TextLoggerHook', by_epoch=True),
        dict(type='TensorboardLoggerHook')
    ])
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = None
resume_from = None
workflow = [('train', 1)]
cudnn_benchmark = True
optimizer = dict(type='Adam', lr=0.0003, betas=[0.9, 0.99])
optimizer_config = dict()
lr_config = dict(policy='CosineAnnealing', min_lr=0.0)
max_epochs = 10
runner = dict(type='EpochBasedRunner', max_epochs=10)
checkpoint_config = dict(by_epoch=True, interval=5)
evaluation = dict(interval=1, save_best='mDice')
work_dir = 'configs/work_dirs'
gpu_ids = range(0, 1)

2022-09-23 20:59:38,287 - MCTrans - INFO - Set random seed to 0, deterministic: False
2022-09-23 21:00:48,728 - MCTrans - INFO - Phase train : Loaded 380 images
2022-09-23 21:01:31,971 - MCTrans - INFO - Phase val : Loaded 203 images
2022-09-23 21:04:52,491 - MCTrans - INFO - Start running, host: wjs@DESKTOP-IVD7SKP, work_dir: Z:\py_item\7_MCTrans-master\MCTrans-master\configs\work_dirs
2022-09-23 21:04:52,495 - MCTrans - INFO - Hooks will be executed in the following order:
before_run:
(VERY_HIGH   ) CosineAnnealingLrUpdaterHook       
(NORMAL      ) CheckpointHook                     
(NORMAL      ) EvalHook                           
(VERY_LOW    ) TextLoggerHook                     
(VERY_LOW    ) TensorboardLoggerHook              
 -------------------- 
before_train_epoch:
(VERY_HIGH   ) CosineAnnealingLrUpdaterHook       
(NORMAL      ) EvalHook                           
(LOW         ) IterTimerHook                      
(VERY_LOW    ) TextLoggerHook                     
(VERY_LOW    ) TensorboardLoggerHook              
 -------------------- 
before_train_iter:
(VERY_HIGH   ) CosineAnnealingLrUpdaterHook       
(NORMAL      ) EvalHook                           
(LOW         ) IterTimerHook                      
 -------------------- 
after_train_iter:
(ABOVE_NORMAL) OptimizerHook                      
(NORMAL      ) CheckpointHook                     
(NORMAL      ) EvalHook                           
(LOW         ) IterTimerHook                      
(VERY_LOW    ) TextLoggerHook                     
(VERY_LOW    ) TensorboardLoggerHook              
 -------------------- 
after_train_epoch:
(NORMAL      ) CheckpointHook                     
(NORMAL      ) EvalHook                           
(VERY_LOW    ) TextLoggerHook                     
(VERY_LOW    ) TensorboardLoggerHook              
 -------------------- 
before_val_epoch:
(LOW         ) IterTimerHook                      
(VERY_LOW    ) TextLoggerHook                     
(VERY_LOW    ) TensorboardLoggerHook              
 -------------------- 
before_val_iter:
(LOW         ) IterTimerHook                      
 -------------------- 
after_val_iter:
(LOW         ) IterTimerHook                      
 -------------------- 
after_val_epoch:
(VERY_LOW    ) TextLoggerHook                     
(VERY_LOW    ) TensorboardLoggerHook              
 -------------------- 
after_run:
(VERY_LOW    ) TextLoggerHook                     
(VERY_LOW    ) TensorboardLoggerHook              
 -------------------- 
2022-09-23 21:04:52,496 - MCTrans - INFO - workflow: [('train', 1)], max: 10 epochs
2022-09-23 21:04:52,498 - MCTrans - INFO - Checkpoints will be saved to Z:\py_item\7_MCTrans-master\MCTrans-master\configs\work_dirs by HardDiskBackend.

**Comment by jscodecode** [comment]
> hi, can you please provide the whole log?

2022-09-23 21:10:50,094 - MCTrans - INFO - Distributed training: False
2022-09-23 21:10:50,534 - MCTrans - INFO - Config:
model = dict(
    type='EncoderDecoder',
    pretrained=None,
    encoder=dict(type='VGG', in_channel=512, init_channels=32, num_blocks=2),
    center=dict(
        type='MCTrans',
        d_model=128,
        nhead=8,
        d_ffn=512,
        dropout=0.1,
        act='relu',
        n_levels=3,
        n_points=4,
        n_sa_layers=6),
    decoder=dict(type='UNetDecoder', in_channels=[32, 64, 128, 128, 128]),
    seg_head=dict(
        type='BasicSegHead',
        in_channels=32,
        num_classes=1,
        post_trans=[
            dict(type='Activations', softmax=True),
            dict(type='AsDiscrete', argmax=True)
        ],
        losses=[
            dict(type='DiceLoss', softmax=True, to_onehot_y=True),
            dict(type='CrossEntropyLoss'),
            dict(type='FLoss')
        ]),
    aux_head=dict(
        type='MCTransAuxHead',
        d_model=128,
        d_ffn=512,
        act='relu',
        num_classes=1,
        in_channles=[32, 64, 128, 128, 128],
        losses=[dict(type='MCTransAuxLoss', sigmoid=True, loss_weight=0.1)]))
dataset_type = 'PanNukeDataset'
patch_size = [512, 512]
keyword = ['img', 'seg_label']
train_transforms = [
    dict(
        type='LoadImage', keys=['img', 'seg_label'], meta_key_postfix='metas'),
    dict(type='AsChannelFirst', keys='img'),
    dict(type='AddChannel', keys='seg_label'),
    dict(
        type='Resize',
        keys=['img', 'seg_label'],
        spatial_size=[512, 512],
        mode=('bilinear', 'nearest')),
    dict(type='ScaleIntensity', keys='img'),
    dict(type='RandMirror'),
    dict(type='ToTensor', keys=['img', 'seg_label'])
]
test_transforms = [
    dict(
        type='LoadImage', keys=['img', 'seg_label'], meta_key_postfix='metas'),
    dict(type='AsChannelFirst', keys='img'),
    dict(type='AddChannel', keys='seg_label'),
    dict(
        type='Resize',
        keys=['img', 'seg_label'],
        spatial_size=[512, 512],
        mode=('bilinear', 'nearest')),
    dict(type='ScaleIntensity', keys='img'),
    dict(type='ToTensor', keys=['img', 'seg_label'])
]
data = dict(
    samples_per_gpu=1,
    workers_per_gpu=1,
    train=dict(
        type='PanNukeDataset',
        img_dirs=[
            'data/medical/pannuke/split-images/BAO SU RI TA LA TU_data',
            'data/medical/pannuke/split-images/DANG YA QING_data'
        ],
        img_suffix='.png',
        label_dirs=[
            'data/medical/pannuke/split-masks/BAO SU RI TA LA TU_label',
            'data/medical/pannuke/split-masks/DANG YA QING_label'
        ],
        label_suffix='-gt.png',
        phase='train',
        transforms=[
            dict(
                type='LoadImage',
                keys=['img', 'seg_label'],
                meta_key_postfix='metas'),
            dict(type='AsChannelFirst', keys='img'),
            dict(type='AddChannel', keys='seg_label'),
            dict(
                type='Resize',
                keys=['img', 'seg_label'],
                spatial_size=[512, 512],
                mode=('bilinear', 'nearest')),
            dict(type='ScaleIntensity', keys='img'),
            dict(type='RandMirror'),
            dict(type='ToTensor', keys=['img', 'seg_label'])
        ]),
    val=dict(
        type='PanNukeDataset',
        img_dirs=['data/medical/pannuke/split-images/BAO SU RI TA LA TU_data'],
        img_suffix='.png',
        label_dirs=[
            'data/medical/pannuke/split-masks/BAO SU RI TA LA TU_label'
        ],
        label_suffix='-gt.png',
        phase='val',
        transforms=[
            dict(
                type='LoadImage',
                keys=['img', 'seg_label'],
                meta_key_postfix='metas'),
            dict(type='AsChannelFirst', keys='img'),
            dict(type='AddChannel', keys='seg_label'),
            dict(
                type='Resize',
                keys=['img', 'seg_label'],
                spatial_size=[512, 512],
                mode=('bilinear', 'nearest')),
            dict(type='ScaleIntensity', keys='img'),
            dict(type='ToTensor', keys=['img', 'seg_label'])
        ]))
log_config = dict(
    interval=10,
    hooks=[
        dict(type='TextLoggerHook', by_epoch=True),
        dict(type='TensorboardLoggerHook')
    ])
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = None
resume_from = None
workflow = [('train', 1)]
cudnn_benchmark = True
optimizer = dict(type='Adam', lr=0.0003, betas=[0.9, 0.99])
optimizer_config = dict()
lr_config = dict(policy='CosineAnnealing', min_lr=0.0)
max_epochs = 10
runner = dict(type='EpochBasedRunner', max_epochs=10)
checkpoint_config = dict(by_epoch=True, interval=5)
evaluation = dict(interval=1, save_best='mDice')
work_dir = 'configs/work_dirs'
gpu_ids = range(0, 1)

2022-09-23 21:10:50,534 - MCTrans - INFO - Set random seed to 0, deterministic: False
2022-09-23 21:10:56,203 - MCTrans - INFO - Phase train : Loaded 380 images
2022-09-23 21:11:00,865 - MCTrans - INFO - Phase val : Loaded 203 images
2022-09-23 21:11:03,875 - MCTrans - INFO - Start running, host: wjs@DESKTOP-IVD7SKP, work_dir: Z:\py_item\7_MCTrans-master\MCTrans-master\configs\work_dirs
2022-09-23 21:11:03,876 - MCTrans - INFO - Hooks will be executed in the following order:
before_run:
(VERY_HIGH   ) CosineAnnealingLrUpdaterHook       
(NORMAL      ) CheckpointHook                     
(NORMAL      ) EvalHook                           
(VERY_LOW    ) TextLoggerHook                     
(VERY_LOW    ) TensorboardLoggerHook              
 -------------------- 
before_train_epoch:
(VERY_HIGH   ) CosineAnnealingLrUpdaterHook       
(NORMAL      ) EvalHook                           
(LOW         ) IterTimerHook                      
(VERY_LOW    ) TextLoggerHook                     
(VERY_LOW    ) TensorboardLoggerHook              
 -------------------- 
before_train_iter:
(VERY_HIGH   ) CosineAnnealingLrUpdaterHook       
(NORMAL      ) EvalHook                           
(LOW         ) IterTimerHook                      
 -------------------- 
after_train_iter:
(ABOVE_NORMAL) OptimizerHook                      
(NORMAL      ) CheckpointHook                     
(NORMAL      ) EvalHook                           
(LOW         ) IterTimerHook                      
(VERY_LOW    ) TextLoggerHook                     
(VERY_LOW    ) TensorboardLoggerHook              
 -------------------- 
after_train_epoch:
(NORMAL      ) CheckpointHook                     
(NORMAL      ) EvalHook                           
(VERY_LOW    ) TextLoggerHook                     
(VERY_LOW    ) TensorboardLoggerHook              
 -------------------- 
before_val_epoch:
(LOW         ) IterTimerHook                      
(VERY_LOW    ) TextLoggerHook                     
(VERY_LOW    ) TensorboardLoggerHook              
 -------------------- 
before_val_iter:
(LOW         ) IterTimerHook                      
 -------------------- 
after_val_iter:
(LOW         ) IterTimerHook                      
 -------------------- 
after_val_epoch:
(VERY_LOW    ) TextLoggerHook                     
(VERY_LOW    ) TensorboardLoggerHook              
 -------------------- 
after_run:
(VERY_LOW    ) TextLoggerHook                     
(VERY_LOW    ) TensorboardLoggerHook              
 -------------------- 
2022-09-23 21:11:03,876 - MCTrans - INFO - workflow: [('train', 1)], max: 10 epochs
2022-09-23 21:11:03,876 - MCTrans - INFO - Checkpoints will be saved to Z:\py_item\7_MCTrans-master\MCTrans-master\configs\work_dirs by HardDiskBackend.

**Comment by JiYuanFeng** [comment]
HI, please set the "num_classes" to 2. (foreground and background classes)

**Comment by jscodecode** [comment]
I konw. Thank you for your help！




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2022年9月24日(星期六) 下午2:09
收件人: ***@***.***&gt;; 
抄送: "ァDê_/~↘松|     ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [JiYuanFeng/MCTrans] &#39;Single-channel predictions not supported&#39; (Issue #25)





 
HI, please set the "num_classes" to 2. (foreground and background classes)
